### PR TITLE
cite potential builds for oc status missing input streams

### LINKED ISSUE #7822: Additional improvements to `oc status` in the build config / input image stream area
Some follow up items from https://github.com/openshift/origin/issues/5077
1) Explore if the outputs of other build configs and match the missing inputs for build configs we've located, and adjust the messages; use a warning message if such build config output exist and tell the user to start those builds, but then use an error message if no such possibilities exist

2) ImageStreamImage nodes currently have edges to ImageStreams; should they have edges to the ImageStream.status.Tag entries they correspond to (either instead of, or in addition to, the edge to the ImageStream.

### PR DESCRIPTION
**Pull request by gabemontero**
Fixes #7822 

@bparees @kargakis @deads2k PTAL if you get the chance.

@bparees I'll try to mark this milestone 1.2.x but I suspect I won't have permission to do so.

quick FYI, in addition to the items noted in the associated issue, my unit testing found a problem with the uniqueNamer / UniqueName code and getting that code executed by the name ResourceName() code instead of using the default reflection code.

Note - I decided to only modify the code to search the output of other BuildConfig's to modify the suggestion for missing inputs to BuildConfigs.

I'm deferring (again, and moving forward pending further evolution of this component) on the other item in this issue:  adding an edge between a ImageStreamImageNode and the corresponding DockerImageRepositoryNode.  My rational simply stems from the additional code cost (DockerImageRepositoryNode exists as a type in the code but not as a node in the graph that is built today) vs. the small amount of code needed to iterate over a single image streams tags.

### CODE REVIEW

**Comment by gabemontero** [comment]
Ah ... my focused go test build missed imageprunner.go ... fixing ...

File: pkg/api/graph/graph.go [hunk 1]
```diff
@@ -122,7 +122,7 @@ var DefaultNamer Namer = namer{}
 func (namer) ResourceName(obj interface{}) string {
 	switch t := obj.(type) {
 	case uniqueNamer:
-		return t.UniqueName()
+		return t.UniqueName().UniqueName()
```
Comment (bparees) [inline pkg/api/graph/graph.go]: wait, what?

File: pkg/build/graph/analysis/bc.go [hunk 2]
```diff
@@ -238,6 +238,16 @@ func FindPendingTags(g osgraph.Graph, f osgraph.Namer) []osgraph.Marker {
 	return markers
 }
 
+// getImageStreamTagSuggestion will return the appropriate marker Suggestion for when a BuildConfig is missing is input ImageStreamTag;  in particular,
```
Comment (bparees) [inline pkg/build/graph/analysis/bc.go]: second is should be its

File: pkg/build/graph/analysis/bc.go [hunk 3]
```diff
@@ -238,6 +238,16 @@ func FindPendingTags(g osgraph.Graph, f osgraph.Namer) []osgraph.Marker {
...
+// it will determine whether or not another BuildConfig can produce the aforementioned ImageStreamTag
+func getImageStreamTagSuggestion(g osgraph.Graph, f osgraph.Namer, tagNode *imagegraph.ImageStreamTagNode, bcNode *buildgraph.BuildConfigNode) osgraph.Suggestion {
+	if _, points := buildPointsToTag(g, tagNode); points {
+		return osgraph.Suggestion(fmt.Sprintf("%s needs to be imported or created by a build of %s.", f.ResourceName(tagNode), f.ResourceName(bcNode)))
```
Comment (bparees) [inline pkg/build/graph/analysis/bc.go]: should handle the case where multiple buildconfigs point to(produce) a particular imagestreamtag?

File: pkg/api/graph/graph.go (hunk 1 above)
Comment (gabemontero) [inline pkg/api/graph/graph.go]: Yeah, I know ... compile error otherwise.  That second UniqueName() could be a String() as well.  I'll admit I left it like this to generate some discussion.

File: pkg/build/graph/analysis/bc.go (hunk 3 above)
Comment (bparees) [inline pkg/build/graph/analysis/bc.go]: also it seems like the suggested text is going to tell the user to run the build that had the problem in the first place, rather than to run the build that would produce the needed imagestreamtag?

Comment (gabemontero) [inline pkg/build/graph/analysis/bc.go]: ahhh ... good catch, I re-factored that toward the end and mucked it up.

Comment (gabemontero) [inline pkg/build/graph/analysis/bc.go]: pushed update to those 2 items

File: pkg/build/graph/analysis/bc.go (hunk 2 above)
Comment (gabemontero) [inline pkg/build/graph/analysis/bc.go]: pushed update

File: pkg/api/graph/graph.go (hunk 1 above)
Comment (bparees) [inline pkg/api/graph/graph.go]: can you elaborate on why the signature change had to be made in the first place?  (I see you said something about reflection issues in the PR summary).

File: pkg/build/graph/analysis/bc.go [hunk 4]
```diff
@@ -238,6 +238,16 @@ func FindPendingTags(g osgraph.Graph, f osgraph.Namer) []osgraph.Marker {
 	return markers
 }
 
+// getImageStreamTagSuggestion will return the appropriate marker Suggestion for when a BuildConfig is missing its input ImageStreamTag;  in particular,
+// it will determine whether or not another BuildConfig can produce the aforementioned ImageStreamTag
+func getImageStreamTagSuggestion(g osgraph.Graph, f osgraph.Namer, tagNode *imagegraph.ImageStreamTagNode, bcNode *buildgraph.BuildConfigNode) osgraph.Suggestion {
+	if inputBCNode, points := buildPointsToTag(g, tagNode); points {
```
Comment (bparees) [inline pkg/build/graph/analysis/bc.go]: i still think this function needs to be updated to return all the BCs that point to the provided IST.  i guess it just returns the first one today?

File: pkg/api/graph/graph.go (hunk 1 above)
Comment (gabemontero) [inline pkg/api/graph/graph.go]: Sure ... items like BuildConfigNode, ImageStreamTagNode, etc. were not passing as instances of the `uniqueNamer` interface, where the `UniqueName()` method has a return of type `UniqueName`.

So I had to pick between getting consistent with a return type of `UniqueName`.  Assuming that the specialized type was created for a good reason, I also based picking `UniqueName` over string on the fact that similar k8s resources (see https://github.com/openshift/origin/blob/master/pkg/api/kubegraph/nodes/types.go) used a return type of `UniqueName`.

Bottom line - had to make things consistent.  We could go back and change the return type to `string` and make the requisite changes in that direction.

File: pkg/build/graph/analysis/bc.go (hunk 4 above)
Comment (gabemontero) [inline pkg/build/graph/analysis/bc.go]: yes, returns just the first one ... ok, i'll update to return the complete list.

File: pkg/api/graph/graph.go (hunk 1 above)
Comment (bparees) [inline pkg/api/graph/graph.go]: i think you should be able to just cast t.UniqueName() to a string?   I guess that's exactly what the second UniqueName() call does.  not sure which is less confusing.

File: pkg/build/graph/analysis/bc.go [hunk 5]
```diff
@@ -238,6 +238,16 @@ func FindPendingTags(g osgraph.Graph, f osgraph.Namer) []osgraph.Marker {
...
+		return osgraph.Suggestion(fmt.Sprintf("%s needs to be imported or created by a build of %s.", f.ResourceName(tagNode), f.ResourceName(inputBCNode)))
+	} else {
```
Comment (kargakis) [inline pkg/build/graph/analysis/bc.go]: drop else

File: pkg/api/graph/graph.go (hunk 1 above)
Comment (kargakis) [inline pkg/api/graph/graph.go]: ouch! not sure about this

Comment (gabemontero) [inline pkg/api/graph/graph.go]: I added a `String()` method for the `UniqueString` type and removed the `UniqueName()` method (which was not being used anywhere other than the local instance I temporarily had) ... seemed a bit cleaner than typecasting to a string.

File: pkg/build/graph/analysis/bc.go (hunk 5 above)
Comment (gabemontero) [inline pkg/build/graph/analysis/bc.go]: ah yes ... that "go-ism" hasn't fully settled into my consciousness yet.

**Comment by gabemontero** [comment]
updates based on comments pushed

File: pkg/build/graph/analysis/bc.go [hunk 6]
```diff
@@ -238,6 +238,19 @@ func FindPendingTags(g osgraph.Graph, f osgraph.Namer) []osgraph.Marker {
 	return markers
 }
 
+// getImageStreamTagSuggestion will return the appropriate marker Suggestion for when a BuildConfig is missing its input ImageStreamTag;  in particular,
+// it will determine whether or not another BuildConfig can produce the aforementioned ImageStreamTag
+func getImageStreamTagSuggestion(g osgraph.Graph, f osgraph.Namer, tagNode *imagegraph.ImageStreamTagNode, bcNode *buildgraph.BuildConfigNode) osgraph.Suggestion {
+	bcs := []string{}
+	if inputBCNode, points := buildPointsToTag(g, tagNode); points {
+		bcs = append(bcs, f.ResourceName(inputBCNode))
```
Comment (bparees) [inline pkg/build/graph/analysis/bc.go]: isn't this still just going to return the first bc that points to tagNode, rather than all of them?

Comment (kargakis) [inline pkg/build/graph/analysis/bc.go]: yes, what needs to change here is buildPointsToTag - also its name can be better, maybe getBuildsForTag

Comment (gabemontero) [inline pkg/build/graph/analysis/bc.go]: duh ... yep, my thought process only got to the first level ... I'll rework the attempt to get all the bc's

on the name, since `git blame` says you originally introduced the `buildPointsToTag` method @kargakis , that certainly simplifies the debate on what to name it to ;-) ... i'll switch to getBuildsForTag and change both the original call to it in `FindPendingTags` and the one I added in `getImageStreamTagSuggestion` for `FindMissingInputStreams`

Comment (gabemontero) [inline pkg/build/graph/analysis/bc.go]: actually, I think I can just move the `PredecessorNodeByEdgeKind` up a level

Comment (gabemontero) [inline pkg/build/graph/analysis/bc.go]: yep, i think that is better ... left `buildPointsToTag` name alone, along with its functionality ... if `FindPendingTags` needs to be reworked, that can be done under a separate PR.

about to push ...

**Comment by gabemontero** [comment]
updates based on comments pushed

**Comment by bparees** [comment]
lgtm.  @kargakis ?

File: pkg/build/graph/analysis/bc.go [hunk 7]
```diff
@@ -238,6 +238,19 @@ func FindPendingTags(g osgraph.Graph, f osgraph.Namer) []osgraph.Marker {
...
 
+// getImageStreamTagSuggestion will return the appropriate marker Suggestion for when a BuildConfig is missing its input ImageStreamTag;  in particular,
+// it will determine whether or not another BuildConfig can produce the aforementioned ImageStreamTag
+func getImageStreamTagSuggestion(g osgraph.Graph, f osgraph.Namer, tagNode *imagegraph.ImageStreamTagNode, bcNode *buildgraph.BuildConfigNode) osgraph.Suggestion {
+	bcs := []string{}
+	for _, bcNode := range g.PredecessorNodesByEdgeKind(tagNode, buildedges.BuildOutputEdgeKind) {
+		bcs = append(bcs, f.ResourceName(bcNode))
+	}
+	if len(bcs) > 0 {
+		return osgraph.Suggestion(fmt.Sprintf("%s needs to be imported or created by a build(s) of %s.", f.ResourceName(tagNode), strings.Join(bcs[:], ",")))
```
Comment (kargakis) [inline pkg/build/graph/analysis/bc.go]: If there is only one build config, we can suggest running `oc start-build`. I would prefer if we separated these cases: no bcs, one bc, multiple bcs.

**Comment by kargakis** [comment]
cc: @stevekuznetsov 

@deads2k for the graph api changes

**Comment by stevekuznetsov** [comment]
> `DockerImageRepositoryNode` exists as a type in the code but not as a node in the graph that is built today

When doing the HPA work this did strike me as a place where we could potentially auto-gen code. Edge builder methods would be hand-written, of course, but the boilerplate definitely doesn't need to be.

**Comment by stevekuznetsov** [comment]
Or perhaps we could have generic nodes that self-identify using reflection and not need boilerplate at all.

File: pkg/build/graph/analysis/bc.go (hunk 7 above)
Comment (deads2k) [inline pkg/build/graph/analysis/bc.go]: > If there is only one build config, we can suggest running oc start-build. I would prefer if we separated these cases: no bcs, one bc, multiple bcs.

Good thought.

File: pkg/image/prune/imagepruner.go [hunk 8]
```diff
@@ -358,7 +358,7 @@ func addImageStreamsToGraph(g graph.Graph, streams *imageapi.ImageStreamList, al
 					continue
 				}
 
-				glog.V(4).Infof("Adding edge (kind=%s) from %q to %q", kind, imageStreamNode.UniqueName.UniqueName(), imageNode.UniqueName.UniqueName())
+				glog.V(4).Infof("Adding edge (kind=%s) from %q to %q", kind, imageStreamNode.UniqueName(), imageNode.String())
```
Comment (deads2k) [inline pkg/image/prune/imagepruner.go]: This change isn't obvious to me.  Why not use the unique name of the `imageNode` as well (like it used to be)?

File: pkg/api/graph/graph.go [hunk 9]
```diff
@@ -41,8 +41,12 @@ func (n UniqueName) UniqueName() string {
 	return string(n)
 }
 
+func (n UniqueName) String() string {
+	return string(n)
+}
+
 type uniqueNamer interface {
-	UniqueName() string
+	UniqueName() UniqueName
```
Comment (deads2k) [inline pkg/api/graph/graph.go]: This change looks good.  I suspect this was the intent and it broke at some point.

File: pkg/build/graph/analysis/bc.go (hunk 7 above)
Comment (gabemontero) [inline pkg/build/graph/analysis/bc.go]: OK, will separate out into no bcs, one bc, multiple bcs.

File: pkg/image/prune/imagepruner.go (hunk 8 above)
Comment (gabemontero) [inline pkg/image/prune/imagepruner.go]: Oversight as I was wrestling with the interface change.  Will switch to `UniqueName()`

**Comment by gabemontero** [comment]
updates from comments of @deads2k and @kargakis pushed.

File: pkg/build/graph/analysis/bc.go [hunk 10]
```diff
@@ -238,6 +238,22 @@ func FindPendingTags(g osgraph.Graph, f osgraph.Namer) []osgraph.Marker {
...
 
+// getImageStreamTagSuggestion will return the appropriate marker Suggestion for when a BuildConfig is missing its input ImageStreamTag;  in particular,
+// it will determine whether or not another BuildConfig can produce the aforementioned ImageStreamTag
+func getImageStreamTagSuggestion(g osgraph.Graph, f osgraph.Namer, tagNode *imagegraph.ImageStreamTagNode, bcNode *buildgraph.BuildConfigNode) osgraph.Suggestion {
+	bcs := []string{}
+	for _, bcNode := range g.PredecessorNodesByEdgeKind(tagNode, buildedges.BuildOutputEdgeKind) {
+		bcs = append(bcs, f.ResourceName(bcNode))
+	}
+	if len(bcs) == 1 {
+		return osgraph.Suggestion(fmt.Sprintf("%s needs to be imported or created by executing `oc start-build %s`.", f.ResourceName(tagNode), strings.Join(bcs[:], ",")))
```
Comment (kargakis) [inline pkg/build/graph/analysis/bc.go]: suggest just the command

Comment (gabemontero) [inline pkg/build/graph/analysis/bc.go]: On Mon, Apr 11, 2016 at 10:13 AM, Michail Kargakis <notifications@github.com
> wrote:

> In pkg/build/graph/analysis/bc.go
> <https://github.com/openshift/origin/pull/8351#discussion_r59210411>:
>
> > @@ -238,6 +238,22 @@ func FindPendingTags(g osgraph.Graph, f osgraph.Namer) []osgraph.Marker {
> >  	return markers
> >  }
> >
> > +// getImageStreamTagSuggestion will return the appropriate marker Suggestion for when a BuildConfig is missing its input ImageStreamTag;  in particular,
> > +// it will determine whether or not another BuildConfig can produce the aforementioned ImageStreamTag
> > +func getImageStreamTagSuggestion(g osgraph.Graph, f osgraph.Namer, tagNode *imagegraph.ImageStreamTagNode, bcNode *buildgraph.BuildConfigNode) osgraph.Suggestion {
> > +	bcs := []string{}
> > +	for _, bcNode := range g.PredecessorNodesByEdgeKind(tagNode, buildedges.BuildOutputEdgeKind) {
> > +		bcs = append(bcs, f.ResourceName(bcNode))
> > +	}
> > +	if len(bcs) == 1 {
> > +		return osgraph.Suggestion(fmt.Sprintf("%s needs to be imported or created by executing `oc start-build %s`.", f.ResourceName(tagNode), strings.Join(bcs[:], ",")))
>
> suggest just the command
>

pushed

File: pkg/build/graph/analysis/bc.go [hunk 11]
```diff
@@ -238,6 +238,22 @@ func FindPendingTags(g osgraph.Graph, f osgraph.Namer) []osgraph.Marker {
...
 
+// getImageStreamTagSuggestion will return the appropriate marker Suggestion for when a BuildConfig is missing its input ImageStreamTag;  in particular,
+// it will determine whether or not another BuildConfig can produce the aforementioned ImageStreamTag
+func getImageStreamTagSuggestion(g osgraph.Graph, f osgraph.Namer, tagNode *imagegraph.ImageStreamTagNode, bcNode *buildgraph.BuildConfigNode) osgraph.Suggestion {
+	bcs := []string{}
+	for _, bcNode := range g.PredecessorNodesByEdgeKind(tagNode, buildedges.BuildOutputEdgeKind) {
+		bcs = append(bcs, f.ResourceName(bcNode))
+	}
+	if len(bcs) == 1 {
+		return osgraph.Suggestion(fmt.Sprintf("%s needs to be created by executing `oc start-build %s`.", f.ResourceName(tagNode), strings.Join(bcs[:], ",")))
```
Comment (kargakis) [inline pkg/build/graph/analysis/bc.go]: See https://github.com/openshift/origin/blob/e7eaff831482b24099caa9aa02e67db5290a442f/pkg/build/graph/analysis/bc.go#L203-l204. Actually, only the command in the case of 1 bc is a suggestion, the rest should be part of the message. So I think getImageSteamTagSuggestion should be getImageStreamTagMarker and return a complete marker instead of just the suggestion.

Comment (kargakis) [inline pkg/build/graph/analysis/bc.go]: We have started using this pattern of building scoped markers elsewhere: https://github.com/openshift/origin/blob/e7eaff831482b24099caa9aa02e67db5290a442f/pkg/route/graph/analysis/analysis.go#L57. Helps a lot in having concise messages+suggestions.

Comment (gabemontero) [inline pkg/build/graph/analysis/bc.go]: OK, I believe I honored the message/suggestion intent with the ImageStreamTag ones.  For ImageStreamImage, after going back and forth, settled on the opinion that those are a bit special case, and left the suggestions as is (and to what we agreed to previously).

Some code restructure which I at least viewed as positive subsequently emerged from the method rename you suggested, so totally cool from that perspective.

PTAL

**Comment by gabemontero** [comment]
pushed update to address last gofmt; also thought some more about it and managed some pruning of ImageStreamImage suggestions along with some `oc import-image --from` pseudo code in the text.

**Comment by gabemontero** [comment]
pushed update to address last gofmt; also thought some more about it and managed some pruning of ImageStreamImage suggestions along with some `oc import-image --from` pseudo code in the text.

**Comment by gabemontero** [comment]
the 1.6 travis passed ... the 1.5 travis failed with govet complaints unrelated to this PR.

File: pkg/api/graph/graph.go (hunk 9 above)
Comment (smarterclayton) [inline pkg/api/graph/graph.go]: There is no point to this change?

Comment (gabemontero) [inline pkg/api/graph/graph.go]: On Tue, Apr 26, 2016 at 1:47 PM, Clayton Coleman <notifications@github.com>
wrote:

> In pkg/api/graph/graph.go
> <https://github.com/openshift/origin/pull/8351#discussion_r61131701>:
>
> >  type uniqueNamer interface {
> > -	UniqueName() string
> > +	UniqueName() UniqueName
>
> There is no point to this change?
>
Actually, this change was needed to properly line up implementations of
this interface and
https://github.com/openshift/origin/blob/master/pkg/api/graph/graph.go#L122-L129
,
My testing found cases where
https://github.com/openshift/origin/blob/master/pkg/api/graph/graph.go#L127
was being called instead of UniqueName()

**Comment by smarterclayton** [comment]
[test]

**Comment by gabemontero** [comment]
@smarterclayton  - FYI - The integration test failure in this PR was the same as with my noInputBC PR.  There issues doing a docker pull on `<registry ip:port>/cache/mysql:pullthrough`

@bparees - given the "mysql" element, is that end-to-end test devexp ?

**Comment by bparees** [comment]
@gabemontero no, the mysql image in this case is the dockerhub image, not our mysql-55-centos7 image.  we're doing a pullthrough which means we're asking the openshift registry to pull the docker.io mysql image on our behalf.  I assume that makes it @pweil-'s team's domain.

**Comment by gabemontero** [comment]
@bparees ah, ok, thx for the explanation.  Perhaps it was just a networking hiccup for the openshift registry trying to access docker.io for this image then, but of course that's just my 2 cents.

**Comment by bparees** [comment]
> Perhaps it was just a networking hiccup for the openshift registry trying to access docker.io for this image then, but of course that's just my 2 cents.

definitely possible, but the registry should have some sort of retry logic if that's the case (not sure if it does, perhaps adding that logic is the fix)

**Comment by gabemontero** [comment]
[test]

**Comment by gabemontero** [comment]
@kargakis @deads2k @bparees bump

File: pkg/build/graph/analysis/bc.go [hunk 12]
```diff
@@ -267,16 +280,34 @@ func validImageStreamImage(imageNode *imagegraph.ImageStreamImageNode, imageStre
...
 		if compareTime.Before(currentTime) {
-			return false, dockerImageReference.ID, "import the image with hexadecimal ID %s into the image stream %s"
+			return osgraph.Suggestion(fmt.Sprintf("`oc import-image %s --from=` where `--from` specifies an image with hexadecimal ID %s", imageStream.GetName(), imageID))
 		}
 
-		return false, dockerImageReference.ID, "a import of the image with hexadecimal ID %s into the image stream %s could be in progress; check again after a couple of minutes"
+		return osgraph.Suggestion(fmt.Sprintf("`oc import-image %s --from=` with hexadecimal ID %s possibly in progress", imageStream.GetName(), imageID))
+
+	}
+	return osgraph.Suggestion(fmt.Sprintf("Possible error occurred with `oc import-image %s --from=` with hexadecimal ID %s; inspect images stream annotations", imageStream.GetName(), imageID))
```
Comment (kargakis) [inline pkg/build/graph/analysis/bc.go]: I would love to have https://github.com/kubernetes/kubernetes/issues/19817 in place for this. Then we could have some thing `" 'oc annotate is/%s' to inspect for possible import error"`

**Comment by kargakis** [comment]
LGTM

**Comment by bparees** [comment]
merging on @kargakis recognizance.
[merge]
